File: python/src/mainlayer_webhooks/events.py
"""
Pydantic models for every Mainlayer webhook event type.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, Literal, Optional, Union

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Base model
# ---------------------------------------------------------------------------


class WebhookEvent(BaseModel):
    """Base shape shared by all Mainlayer webhook events."""

    id: str = Field(..., description="Unique event identifier.")
    type: str = Field(..., description="Event type (e.g. 'payment.completed').")
    data: Dict[str, Any] = Field(..., description="Event-specific payload.")
    created_at: datetime = Field(..., description="ISO 8601 UTC timestamp.")

    model_config = {"extra": "allow"}


# ---------------------------------------------------------------------------
# Payment events
# ---------------------------------------------------------------------------


class PaymentData(BaseModel):
    payment_id: str
    amount: int = Field(..., description="Amount in the smallest currency unit (e.g. cents).")
    currency: str = Field(..., description="Three-letter ISO 4217 currency code.")
    customer_id: str
    description: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None

    model_config = {"extra": "allow"}


class PaymentCompletedData(PaymentData):
    status: Literal["completed"]
    completed_at: datetime


class PaymentRefundedData(PaymentData):
    refund_id: str
    refund_amount: int
    reason: Optional[str] = None
    refunded_at: datetime


class PaymentCompletedEvent(WebhookEvent):
    type: Literal["payment.completed"]
    data: PaymentCompletedData  # type: ignore[assignment]


class PaymentRefundedEvent(WebhookEvent):
    type: Literal["payment.refunded"]
    data: PaymentRefundedData  # type: ignore[assignment]


# ---------------------------------------------------------------------------
# Subscription events
# ---------------------------------------------------------------------------


class SubscriptionData(BaseModel):
    subscription_id: str
    customer_id: str
    plan_id: str
    plan_name: str
    amount: int
    currency: str
    interval: Literal["monthly", "annual", "weekly"]
    current_period_start: datetime
    current_period_end: datetime
    metadata: Optional[Dict[str, str]] = None

    model_config = {"extra": "allow"}


class SubscriptionCreatedData(SubscriptionData):
    status: Literal["active"]
    trial_end: Optional[datetime] = None


class SubscriptionRenewedData(SubscriptionData):
    status: Literal["active"]
    renewed_at: datetime
    next_renewal: datetime


class SubscriptionCancelledData(SubscriptionData):
    status: Literal["cancelled"]
    cancelled_at: datetime
    cancellation_reason: Optional[str] = None
    ends_at: datetime


class SubscriptionCreatedEvent(WebhookEvent):
    type: Literal["subscription.created"]
    data: SubscriptionCreatedData  # type: ignore[assignment]


class SubscriptionRenewedEvent(WebhookEvent):
    type: Literal["subscription.renewed"]
    data: SubscriptionRenewedData  # type: ignore[assignment]


class SubscriptionCancelledEvent(WebhookEvent):
    type: Literal["subscription.cancelled"]
    data: SubscriptionCancelledData  # type: ignore[assignment]


# ---------------------------------------------------------------------------
# Entitlement events
# ---------------------------------------------------------------------------


class EntitlementData(BaseModel):
    entitlement_id: str
    customer_id: str
    feature_key: str
    feature_name: str
    subscription_id: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None

    model_config = {"extra": "allow"}


class EntitlementExpiredData(EntitlementData):
    expired_at: datetime
    expires_at: datetime


class EntitlementExpiredEvent(WebhookEvent):
    type: Literal["entitlement.expired"]
    data: EntitlementExpiredData  # type: ignore[assignment]


# ---------------------------------------------------------------------------
# Discriminated union
# ---------------------------------------------------------------------------

TypedWebhookEvent = Union[
    PaymentCompletedEvent,
    PaymentRefundedEvent,
    SubscriptionCreatedEvent,
    SubscriptionRenewedEvent,
    SubscriptionCancelledEvent,
    EntitlementExpiredEvent,
]

EVENT_TYPE_MAP: Dict[str, type] = {
    "payment.completed": PaymentCompletedEvent,
    "payment.refunded": PaymentRefundedEvent,
    "subscription.created": SubscriptionCreatedEvent,
    "subscription.renewed": SubscriptionRenewedEvent,
    "subscription.cancelled": SubscriptionCancelledEvent,
    "entitlement.expired": EntitlementExpiredEvent,
}

File: typescript/src/events.ts
/**
 * Typed event interfaces for all Mainlayer webhook event types.
 */

export type WebhookEventType =
  | 'payment.completed'
  | 'payment.refunded'
  | 'subscription.created'
  | 'subscription.renewed'
  | 'subscription.cancelled'
  | 'entitlement.expired';

/**
 * Base webhook event shape shared by all event types.
 */
export interface WebhookEvent {
  id: string;
  type: WebhookEventType;
  data: Record<string, unknown>;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Payment events
// ---------------------------------------------------------------------------

export interface PaymentData {
  [key: string]: unknown;
  payment_id: string;
  amount: number;
  currency: string;
  customer_id: string;
  description?: string;
  metadata?: Record<string, string>;
}

export interface PaymentCompletedData extends PaymentData {
  status: 'completed';
  completed_at: string;
}

export interface PaymentRefundedData extends PaymentData {
  refund_id: string;
  refund_amount: number;
  reason?: string;
  refunded_at: string;
}

export interface PaymentCompletedEvent extends WebhookEvent {
  type: 'payment.completed';
  data: PaymentCompletedData;
}

export interface PaymentRefundedEvent extends WebhookEvent {
  type: 'payment.refunded';
  data: PaymentRefundedData;
}

// ---------------------------------------------------------------------------
// Subscription events
// ---------------------------------------------------------------------------

export interface SubscriptionData {
  [key: string]: unknown;
  subscription_id: string;
  customer_id: string;
  plan_id: string;
  plan_name: string;
  amount: number;
  currency: string;
  interval: 'monthly' | 'annual' | 'weekly';
  current_period_start: string;
  current_period_end: string;
  metadata?: Record<string, string>;
}

export interface SubscriptionCreatedData extends SubscriptionData {
  status: 'active';
  trial_end?: string;
}

export interface SubscriptionRenewedData extends SubscriptionData {
  status: 'active';
  renewed_at: string;
  next_renewal: string;
}

export interface SubscriptionCancelledData extends SubscriptionData {
  status: 'cancelled';
  cancelled_at: string;
  cancellation_reason?: string;
  ends_at: string;
}

export interface SubscriptionCreatedEvent extends WebhookEvent {
  type: 'subscription.created';
  data: SubscriptionCreatedData;
}

export interface SubscriptionRenewedEvent extends WebhookEvent {
  type: 'subscription.renewed';
  data: SubscriptionRenewedData;
}

export interface SubscriptionCancelledEvent extends WebhookEvent {
  type: 'subscription.cancelled';
  data: SubscriptionCancelledData;
}

// ---------------------------------------------------------------------------
// Entitlement events
// ---------------------------------------------------------------------------

export interface EntitlementData {
  [key: string]: unknown;
  entitlement_id: string;
  customer_id: string;
  feature_key: string;
  feature_name: string;
  subscription_id?: string;
  metadata?: Record<string, string>;
}

export interface EntitlementExpiredData extends EntitlementData {
  expired_at: string;
  expires_at: string;
}

export interface EntitlementExpiredEvent extends WebhookEvent {
  type: 'entitlement.expired';
  data: EntitlementExpiredData;
}

// ---------------------------------------------------------------------------
// Discriminated union helper
// ---------------------------------------------------------------------------

export type TypedWebhookEvent =
  | PaymentCompletedEvent
  | PaymentRefundedEvent
  | SubscriptionCreatedEvent
  | SubscriptionRenewedEvent
  | SubscriptionCancelledEvent
  | EntitlementExpiredEvent;
